idaes/apps/uncertainty_propagation/sens.py

```diff
@@ -0,0 +1,750 @@
+            shutil.move("col_row.nl","./dsdp/")
+            shutil.move("col_row.col","./dsdp/")
+            shutil.move("col_row.row","./dsdp/")
+            shutil.move("conorder.txt","./dsdp/")
+            shutil.move("delta_p.out","./dsdp/")
+            shutil.move("dot_out.out","./dsdp/")
+            shutil.move("timings_dot_driver_dsdp.txt", "./dsdp/")
+            shutil.move("timings_k_aug_dsdp.txt", "./dsdp/")
+        except OSError:
+            pass
+
+    return m
+
+def get_dsdp(model, theta_names, theta, var_dic={},tee=False, solver_options=None):
+    """This function calculates gradient vector of the (decision variables, parameters)
+        with respect to the paramerters (theta_names). 
+
+    e.g) min f:  p1*x1+ p2*(x2^2) + p1*p2
+         s.t  c1: x1 + x2 = p1
+              c2: x2 + x3 = p2
+              0 <= x1, x2, x3 <= 10
+              p1 = 10
+              p2 = 5
+    the function retuns dx/dp and dp/dp, and colum orders.
+    
+    The following terms are used to define the output dimensions:
+    Ncon   = number of constraints
+    Nvar   = number of variables (Nx + Ntheta)
+    Nx     = the numer of decision (primal) variables
+    Ntheta = number of uncertain parameters.
+
+    Parameters
+    ----------
+    model: Pyomo ConcreteModel
+        model should includes an objective function
+    theta_names: list of strings
+        List of Var names
+    theta: dict
+        Estimated parameters e.g) from parmest
+    tee: bool, optional
+        Indicates that ef solver output should be teed
+    solver_options: dict, optional
+        Provides options to the solver (also the name of an attribute)
+    var_dic: dictionary
+        If any original variable contains "'", need an auxiliary dictionary 
+        with keys theta_names without "'", values with "'".
+        e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']",
+                       'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"}
+
+    Returns
+    -------
+    dsdp: scipy.sparse.csr.csr_matrix
+        Ntheta by Nvar size sparse matrix. A Jacobian matrix of the (decision variables, parameters) 
+        with respect to paramerters (=theta_name). number of rows = len(theta_name), number of columns= len(col)
+    col: list
+        List of variable names
+    """
+    m = model.clone()
+    original_Param = []
+    perturbed_Param = []
+    m.extra = ConstraintList()
+    kk = 0
+    if var_dic == {}:
+        for i in theta_names:
+            var_dic[i] = i
+    for v in theta_names:
+        v_tmp = str(kk)
+        original_param_object = Param(initialize=theta[v], mutable=True)
+        perturbed_param_object = Param(initialize=theta[v])
+        m.add_component("original_"+v_tmp, original_param_object)
+        m.add_component("perturbed_"+v_tmp, perturbed_param_object)
+        m.extra.add(eval('m.'+var_dic[v]) - eval('m.original_'+v_tmp) == 0 )
```

Comment:
1. Can you share the variable names in the example? I will check if I can run it. 

2. I tested with 

```python
for i, name in enumerate(theta_names):
    orig_param = Param(initialize=theta[name], mutable=True)
    ptb_param = Param(initialize=theta[name])
    m.add_component("original_%s" % i, orig_param)
    m.add_component("perturbed_%s" % i,ptb_param)
    cuid = ComponentUID(name)
    var = cuid.find_component_on(m)
    m.extra.add(var - orig_param == 0)

m_k_aug_dsdp = sensitivity_calculation('k_aug',m,orig_param,ptb_param, tee)
```

but it fails. but it fails. Do I need to update other parts as well?
```python
WARNING: Reassigning the non-component attribute _paramList on block
    (model)._SENSITIVITY_TOOLBOX_DATA with a new Component with type <class
    'pyomo.core.base.param.SimpleParam'>. This is usually indicative of a
    modelling error. To avoid this warning, explicitly delete the attribute:
        del _SENSITIVITY_TOOLBOX_DATA._paramList
Traceback (most recent call last):
  File "uncertainty_propagation_rooney.py", line 35, in <module>
    results =  quantify_propagate_uncertainty(rooney_biegler_model,rooney_biegler_model_opt, data, variable_name, SSE)
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/idaes/apps/uncertainty_propagation/uncertainties.py", line 92, in quantify_propagate_uncertainty
    propagate_results  =  propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee)
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/idaes/apps/uncertainty_propagation/uncertainties.py", line 153, in propagate_uncertainty
    dsdp, col = get_dsdp(model, theta_names, theta, var_dic,tee)      
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/idaes/apps/uncertainty_propagation/sens.py", line 278, in get_dsdp
    m_k_aug_dsdp = sensitivity_calculation('k_aug',m,orig_param,ptb_param, tee)
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/idaes/apps/uncertainty_propagation/sens.py", line 153, in sensitivity_calculation
    sens.setup_sensitivity(paramList)
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/idaes/apps/uncertainty_propagation/sens.py", line 679, in setup_sensitivity
    block._paramList = paramList
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/Pyomo-5.7.2a0-py3.8-linux-x86_64.egg/pyomo/core/base/block.py", line 651, in __setattr__
    self.add_component(name, val)
  File "/home/janghopark2/anaconda3/envs/idaes_kaug_v222/lib/python3.8/site-packages/Pyomo-5.7.2a0-py3.8-linux-x86_64.egg/pyomo/core/base/block.py", line 951, in add_component
    raise RuntimeError("""%s
RuntimeError: Re-assigning the component 'original_1' from block 'unknown' to
block '_SENSITIVITY_TOOLBOX_DATA' as '_paramList'.
```

Reply:
updated. `eval`s are removed. Thanks!